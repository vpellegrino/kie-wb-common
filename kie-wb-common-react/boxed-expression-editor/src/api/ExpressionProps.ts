/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { LogicType } from "./LogicType";
import { DataType } from "./DataType";

export interface ExpressionProps {
  /** Expression name (which, in DMN world, is equal to the Decision node's name) */
  name: string;
  /** Expression data type */
  dataType: DataType;
  /** Optional callback executed to update expression's name and data type */
  onUpdatingNameAndDataType?: (updatedName: string, updatedDataType: DataType) => void;
  /** Logic type should not be defined at this stage */
  logicType?: LogicType;
}

export interface LiteralExpressionProps extends ExpressionProps {
  /** Logic type must be LiteralExpression */
  logicType: LogicType.LiteralExpression;
  /** Optional content to display for this literal expression */
  content?: string;
  /** True to have no header for this specific literal expression */
  isHeadless?: boolean;
}
