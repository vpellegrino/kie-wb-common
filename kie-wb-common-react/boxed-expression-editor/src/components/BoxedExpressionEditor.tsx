/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import '@patternfly/react-core/dist/styles/base.css';
import {I18nDictionariesProvider} from "@kogito-tooling/i18n/dist/react-components";
import {ExpressionContainer} from "./ExpressionContainer/ExpressionContainer";
import {
  boxedExpressionEditorDictionaries,
  BoxedExpressionEditorI18nContext,
  boxedExpressionEditorI18nDefaults
} from "../i18n";

const BoxedExpressionEditor: React.FunctionComponent = () => (
  <I18nDictionariesProvider
    defaults={boxedExpressionEditorI18nDefaults}
    dictionaries={boxedExpressionEditorDictionaries}
    initialLocale={navigator.language}
    ctx={BoxedExpressionEditorI18nContext}
  >
    <ExpressionContainer name="TEST"/>
  </I18nDictionariesProvider>
);

export {BoxedExpressionEditor};
