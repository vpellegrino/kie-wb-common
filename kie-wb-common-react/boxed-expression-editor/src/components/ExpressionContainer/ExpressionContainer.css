/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.expression-container .expression-container-box {
  min-width: 170px;
  min-height: 50px;
  border-style: ridge;
  border-color: #39A5DC;
  display: flex;
  justify-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
}

.expression-container .context-menu-container {
  position: absolute;
}

.expression-container > .expression-container-box.logic-type-not-present {
  cursor: pointer;
}

.expression-container > .expression-container-box.logic-type-selected {
  cursor: default;
  padding: 10px;
}

.expression-container > .expression-container-box,
.expression-container .expression-type {
  font-size: smaller;
  font-style: italic;
  color: gray;
}
