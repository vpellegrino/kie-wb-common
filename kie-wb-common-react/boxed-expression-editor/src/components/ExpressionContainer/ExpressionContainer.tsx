/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { useCallback, useMemo, useState } from "react";
import * as _ from "lodash";
import "./ExpressionContainer.css";
import { useBoxedExpressionEditorI18n } from "../../i18n";
import {
  Dropdown,
  DropdownItem,
  KebabToggle,
  SimpleList,
  SimpleListItem,
  SimpleListItemProps,
} from "@patternfly/react-core";
import { PopoverMenu } from "../PopoverMenu";
import { ExpressionProps, LiteralExpressionProps, LogicType } from "../../api";
import { LiteralExpression } from "../LiteralExpression";

export interface ExpressionContainerProps {
  /** Expression properties */
  selectedExpression: ExpressionProps;
}

export const ExpressionContainer: ({ selectedExpression }: ExpressionContainerProps) => JSX.Element = (
  props: ExpressionContainerProps
) => {
  const { i18n } = useBoxedExpressionEditorI18n();

  const [logicTypeIsPresent, setLogicTypeSelected] = useState(
    !_.isEmpty(props.selectedExpression.logicType) || props.selectedExpression.logicType === LogicType.Undefined
  );
  const [actionDropdownIsOpen, setActionDropDownOpen] = useState(false);
  const [selectedExpression, setSelectedExpression] = useState(props.selectedExpression);

  const onLogicTypeSelect = useCallback(
    (currentItem: React.RefObject<HTMLButtonElement>, currentItemProps: SimpleListItemProps) => {
      setLogicTypeSelected(true);
      const selectedLogicType = currentItemProps.children as LogicType;
      setSelectedExpression((previousSelectedExpression: ExpressionProps) => ({
        ...previousSelectedExpression,
        logicType: selectedLogicType,
      }));
    },
    []
  );

  const executeClearAction = useCallback(() => {
    setLogicTypeSelected(false);
    setSelectedExpression((previousSelectedExpression: ExpressionProps) => ({
      name: previousSelectedExpression.name,
      dataType: previousSelectedExpression.dataType,
      logicType: LogicType.Undefined,
    }));
  }, []);

  const onDropdownToggle = useCallback((isOpen) => {
    return setActionDropDownOpen(isOpen);
  }, []);

  const onExpressionActionDropdownSelect = useCallback(
    (actionDropdownIsOpen) => setActionDropDownOpen(!actionDropdownIsOpen),
    []
  );

  const renderExpressionActionsDropdown = useCallback(() => {
    return (
      <Dropdown
        onSelect={onExpressionActionDropdownSelect}
        toggle={<KebabToggle onToggle={onDropdownToggle} className="expression-actions-toggle" />}
        isOpen={actionDropdownIsOpen}
        isPlain
        dropdownItems={[
          <DropdownItem key="clear" onClick={executeClearAction} isDisabled={!logicTypeIsPresent}>
            {i18n.clear}
          </DropdownItem>,
        ]}
      />
    );
  }, [
    i18n.clear,
    onExpressionActionDropdownSelect,
    onDropdownToggle,
    actionDropdownIsOpen,
    logicTypeIsPresent,
    executeClearAction,
  ]);

  const getLogicTypesWithoutUndefined = useCallback(() => {
    return Object.values(LogicType).filter((logicType) => logicType !== LogicType.Undefined);
  }, []);

  const renderLogicTypeItems = useCallback(() => {
    return _.map(getLogicTypesWithoutUndefined(), (key) => <SimpleListItem key={key}>{key}</SimpleListItem>);
  }, [getLogicTypesWithoutUndefined]);

  const getLogicSelectionArrowPlacement = useCallback(
    () => document.querySelector(".expression-container-box")! as HTMLElement,
    []
  );

  const buildLogicSelectorMenu = useCallback(() => {
    return (
      <PopoverMenu
        title={i18n.selectLogicType}
        arrowPlacement={getLogicSelectionArrowPlacement}
        body={<SimpleList onSelect={onLogicTypeSelect}>{renderLogicTypeItems()}</SimpleList>}
      />
    );
  }, [i18n.selectLogicType, getLogicSelectionArrowPlacement, onLogicTypeSelect, renderLogicTypeItems]);

  const updateNameAndDataType = useCallback((updatedName, updatedDataType) => {
    setSelectedExpression((previousSelectedExpression: ExpressionProps) => ({
      ...previousSelectedExpression,
      name: updatedName,
      dataType: updatedDataType,
    }));
  }, []);

  const renderSelectedExpression = useMemo(() => {
    switch (selectedExpression.logicType) {
      case LogicType.LiteralExpression:
        return (
          <LiteralExpression
            onUpdatingNameAndDataType={updateNameAndDataType}
            {...(selectedExpression as LiteralExpressionProps)}
          />
        );
      case LogicType.Context:
      case LogicType.DecisionTable:
      case LogicType.Relation:
      case LogicType.Function:
      case LogicType.Invocation:
      case LogicType.List:
      default:
        return selectedExpression.logicType;
    }
  }, [selectedExpression, updateNameAndDataType]);

  return (
    <div className="expression-container">
      <span className="expression-title">{selectedExpression.name}</span>
      <span className="expression-type">({selectedExpression.logicType || LogicType.Undefined})</span>
      <span className="expression-actions">{renderExpressionActionsDropdown()}</span>

      <div
        className={`expression-container-box ${logicTypeIsPresent ? "logic-type-selected" : "logic-type-not-present"}`}
      >
        {selectedExpression.logicType ? renderSelectedExpression : i18n.selectExpression}
      </div>

      {!logicTypeIsPresent ? buildLogicSelectorMenu() : null}
    </div>
  );
};
