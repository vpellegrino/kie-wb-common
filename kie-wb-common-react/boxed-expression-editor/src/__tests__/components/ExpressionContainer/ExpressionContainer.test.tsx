/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ExpressionContainer } from "../../../components/ExpressionContainer";
import { render } from "@testing-library/react";
import * as React from "react";
import { usingTestingBoxedExpressionI18nContext } from "../test-utils";
import { DataType, LogicType } from "../../../api";
import { act } from "react-dom/test-utils";

jest.useFakeTimers();
const flushPromises = () => new Promise((resolve) => process.nextTick(resolve));

describe("ExpressionContainer tests", () => {
  test("should render ExpressionContainer component", () => {
    const expression = { name: "Test", dataType: DataType.Undefined };

    const { container } = render(
      usingTestingBoxedExpressionI18nContext(<ExpressionContainer selectedExpression={expression} />).wrapper
    );

    expect(container).toMatchSnapshot();
  });

  test("should render expression title, when name prop is passed", () => {
    const expressionTitle = "Test";
    const expression = { name: expressionTitle, dataType: DataType.Undefined };

    const { container } = render(
      usingTestingBoxedExpressionI18nContext(<ExpressionContainer selectedExpression={expression} />).wrapper
    );
    expect(container.querySelector(".expression-title")).toBeTruthy();
    expect(container.querySelector(".expression-title")!.innerHTML).toBe(expressionTitle);
  });

  test("should render expression type, when type prop is passed", () => {
    const expression = { name: "Test", logicType: LogicType.Context, dataType: DataType.Undefined };
    const { container } = render(
      usingTestingBoxedExpressionI18nContext(<ExpressionContainer selectedExpression={expression} />).wrapper
    );

    expect(container.querySelector(".expression-type")).toBeTruthy();
    expect(container.querySelector(".expression-type")!.innerHTML).toBe("(" + LogicType.Context + ")");
  });

  test("should render expression type as undefined, when type prop is not passed", () => {
    const expression = { name: "Test", dataType: DataType.Undefined };

    const { container } = render(
      usingTestingBoxedExpressionI18nContext(<ExpressionContainer selectedExpression={expression} />).wrapper
    );
    expect(container.querySelector(".expression-type")).toBeTruthy();
    expect(container.querySelector(".expression-type")!.innerHTML).toBe("(&lt;Undefined&gt;)");
  });

  describe("Expression Actions dropdown", () => {
    test("should have the clear action disabled on startup", async () => {
      const expression = { name: "Test", dataType: DataType.Undefined };

      const { container } = render(
        usingTestingBoxedExpressionI18nContext(<ExpressionContainer selectedExpression={expression} />).wrapper
      );

      await triggerContextMenu(container, ".expression-container-box");

      expect(container.querySelector(".context-menu-container button.pf-m-disabled")).toBeTruthy();
      expect(container.querySelector(".context-menu-container button.pf-m-disabled")!.innerHTML).toBe("Clear");
    });

    test("should have the clear action enabled, when logic type is selected", async () => {
      const expression = { name: "Test", logicType: LogicType.LiteralExpression, dataType: DataType.Undefined };

      const { container } = render(
        usingTestingBoxedExpressionI18nContext(<ExpressionContainer selectedExpression={expression} />).wrapper
      );

      await triggerContextMenu(container, ".expression-container-box");

      expect(container.querySelector(".context-menu-container button.pf-m-disabled")).toBeFalsy();
      expect(container.querySelector(".context-menu-container button")).toBeTruthy();
      expect(container.querySelector(".context-menu-container button")!.innerHTML).toBe("Clear");
    });
  });

  describe("Logic type selection", () => {
    test("should show the pre-selection, when logic type prop is passed", () => {
      const expression = { name: "Test", logicType: LogicType.Context, dataType: DataType.Undefined };

      const { container } = render(
        usingTestingBoxedExpressionI18nContext(<ExpressionContainer selectedExpression={expression} />).wrapper
      );

      expect(container.querySelector(".expression-container-box")).toBeTruthy();
      expect(container.querySelector(".expression-container-box")!.innerHTML).toBe(expression.logicType);
    });

    test("should reset the selection, when logic type is selected and clear button gets clicked", async () => {
      const expression = { name: "Test", logicType: LogicType.LiteralExpression, dataType: DataType.Undefined };

      const { container } = render(
        usingTestingBoxedExpressionI18nContext(<ExpressionContainer selectedExpression={expression} />).wrapper
      );

      await triggerContextMenu(container, ".expression-container-box");

      act(() => {
        const clearButtonElement = container.querySelector(".context-menu-container button")!;
        const clearButton = clearButtonElement as HTMLButtonElement;
        clearButton.click();
      });

      expect(container.querySelector(".expression-container-box")).toBeTruthy();
      expect(container.querySelector(".expression-container-box")!.innerHTML).not.toBe(expression.logicType);
    });
  });
});

const triggerContextMenu = async (container: HTMLElement, selector: string) => {
  await act(async () => {
    const element = container.querySelector(selector)!;

    element.dispatchEvent(
      new MouseEvent("contextmenu", {
        bubbles: true,
        cancelable: false,
        view: window,
        button: 2,
        buttons: 0,
        clientX: element.getBoundingClientRect().x,
        clientY: element.getBoundingClientRect().y,
      })
    );

    await flushPromises();
    jest.runAllTimers();
  });
};
