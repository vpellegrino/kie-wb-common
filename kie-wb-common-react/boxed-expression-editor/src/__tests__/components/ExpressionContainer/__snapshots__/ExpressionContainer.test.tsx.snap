// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpressionContainer tests should render ExpressionContainer component 1`] = `
<div>
  <div
    class="expression-container"
  >
    <span
      class="expression-title"
    >
      Test
    </span>
    <span
      class="expression-type"
    >
      (
      &lt;Undefined&gt;
      )
    </span>
    <div
      class="expression-container-box logic-type-not-present"
    >
      Select expression
    </div>
  </div>
</div>
`;
